server/testCases/src/types.ts

```diff
@@ -101,3 +105,24 @@ export type Challenge = {
    */
   restrictions: string[];
 };
+
+/**
+ * Return type for a successful test case execution
+ */
+export type TestCaseSuccess = {
+  result: TestCaseResult;
+  error?: undefined;
+};
+
+/**
+ * Return type for a failed test case execution
+ */
+export type TestCaseError = {
+  error: string;
+  result?: undefined;
+};
+
+/**
+ * Discriminated union type for test case execution results
```

Comment:
i dont think this strictly adheres to what a discriminated union is because the discriminant has to be in every variation of the type. Here is an example:

export type TestCaseSuccess = {
  passed: true;
  result: TestCaseResult;
};

export type TestCaseError = {
  passed: false;
  error: string;
};

In this case, `passed` acts as the discriminant for a type union that is a bit cleaner